packages/core/crates/core-types/src/channels.rs

```diff
@@ -204,195 +214,369 @@ impl Default for Ticket {
+        own_address: Address,
+        counterparty: Address,
+        amount: Balance,
+        index: U256,
+        index_offset: U256,
+        win_prob: f64,
+        channel_epoch: U256,
+    ) -> Result<()> {
+        if own_address.eq(&counterparty) {
+            return Err(CoreTypesError::InvalidInputData(
+                "Source and destination must be different".into(),
+            ));
+        }
+
+        if amount.balance_type().ne(&BalanceType::HOPR) {
+            return Err(CoreTypesError::InvalidInputData(
+                "Tickets can only have HOPR balance".into(),
+            ));
+        }
+
+        if amount.value().ge(&10u128.pow(25).into()) {
+            return Err(CoreTypesError::InvalidInputData(
+                "Tickets may not have more than 1% of total supply".into(),
+            ));
+        }
+
+        if index.gt(&(1u64 << 48).into()) {
+            return Err(CoreTypesError::InvalidInputData(
+                "Cannot hold ticket indices larger than 2^48".into(),
+            ));
+        }
+
+        if index_offset.gt(&(1u64 << 32).into()) {
+            return Err(CoreTypesError::InvalidInputData(
+                "Cannot hold ticket index offsets larger than 2^32".into(),
+            ));
+        }
+
+        if channel_epoch.gt(&(1u64 << 24).into()) {
+            return Err(CoreTypesError::InvalidInputData(
+                "Cannot hold channel epoch larger than 2^24".into(),
+            ));
+        }
+
+        if win_prob < 0.0 {
+            return Err(CoreTypesError::InvalidInputData(
+                "Cannot use negative winning ptobability".into(),
+            ));
+        }
+
+        if win_prob > 1.0 {
+            return Err(CoreTypesError::InvalidInputData(
+                "Cannot use winning ptobabilities larger than 100%".into(),
+            ));
+        }
+
+        Ok(())
     }
 
-    pub fn set_challenge(&mut self, challenge: EthereumChallenge, signing_key: &ChainKeypair) {
+    /// Add the challenge property and signs the finished ticket afterwards
+    pub fn set_challenge(&mut self, challenge: EthereumChallenge, signing_key: &ChainKeypair, domain_separator: &Hash) {
         self.challenge = challenge;
-        self.sign(signing_key);
+        self.sign(signing_key, domain_separator);
+    }
+
+    /// Encode winning probability such that it can get used in the smart
+    /// contract
+    pub fn encoded_win_prob(&self) -> u64 {
+        let mut encoded = [0u8; 8];
+        encoded[1..].copy_from_slice(&f64_to_win_prob(self.win_prob).unwrap());
+
+        u64::from_be_bytes(encoded)
+    }
+
+    /// Serializes the ticket with or without signature
+    /// 
+    /// Signing requires hashing which requires serialization without signature.
+    /// Transferring ticket requires serialization with signature attached.
+    fn to_bytes_internal(&self, with_signature: bool) -> Result<Vec<u8>> {
+        let mut ret = Vec::<u8>::with_capacity(if with_signature {
```

Comment:
This is just capacity, not length, so it is ok to overallocate by 64 bytes. So you can just do `Vec::<u8>::with_capacity(Self::SIZE);`

Reply:
I find the stricter way more intuitive